src/credentials/DeviceAttestationConstructor.cpp

```diff
@@ -0,0 +1,173 @@
+            }
+            else
+            {
+                // check that vendorId and profileNum match in every Vendor Reserved entry
+                VerifyOrReturnError(currentVendorId == vendorId && currentProfileNum == profileNum,
+                                    CHIP_ERROR_IM_MALFORMED_COMMAND_DATA_ELEMENT);
+            }
+
+            ByteSpan vendorReservedEntry;
+            ReturnErrorOnFailure(tlvReader.Get(vendorReservedEntry));
+            VerifyOrReturnError(vendorReservedIdx < vendorReservedArraySize, CHIP_ERROR_NO_MEMORY);
+            vendorReservedArray[vendorReservedIdx++] = vendorReservedEntry;
+        }
+        else
+        {
+            return CHIP_ERROR_IM_MALFORMED_COMMAND_DATA_ELEMENT;
+        }
+    }
+
+    VerifyOrReturnError(error == CHIP_END_OF_TLV, error);
+    VerifyOrReturnError(certificationDeclarationExists && attestationNonceExists && timestampExists,
+                        CHIP_ERROR_MISSING_TLV_ELEMENT);
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ConstructAttestationElements(const ByteSpan & certificationDeclaration, const ByteSpan & attestationNonce,
+                                        uint32_t timestamp, const ByteSpan & firmwareInfo, ByteSpan * vendorReservedArray,
+                                        size_t vendorReservedArraySize, uint16_t vendorId, uint16_t profileNum,
+                                        MutableByteSpan & attestationElements)
+{
+    TLV::TLVWriter tlvWriter;
+    TLV::TLVType outerContainerType = TLV::kTLVType_NotSpecified;
+
+    VerifyOrReturnError(!certificationDeclaration.empty() && !attestationNonce.empty(), CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(attestationNonce.size() == 32, CHIP_ERROR_INVALID_MESSAGE_LENGTH);
+    if (vendorReservedArraySize != 0)
+    {
+        VerifyOrReturnError(vendorReservedArray != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    }
+
+    tlvWriter.Init(attestationElements.data(), static_cast<uint32_t>(attestationElements.size()));
+    outerContainerType = TLV::kTLVType_NotSpecified;
+    ReturnErrorOnFailure(tlvWriter.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
+    ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(1), certificationDeclaration));
+    ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(2), attestationNonce));
+    ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(3), timestamp));
+    if (!firmwareInfo.empty())
+    {
+        ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(4), firmwareInfo));
+    }
+
+    uint8_t vendorTagNum = 1;
+    for (size_t vendorReservedIdx = 0; vendorReservedIdx < vendorReservedArraySize; ++vendorReservedIdx)
+    {
+        if (!vendorReservedArray[vendorReservedIdx].empty())
+        {
+            ReturnErrorOnFailure(
+                tlvWriter.Put(TLV::ProfileTag(vendorId, profileNum, vendorTagNum), vendorReservedArray[vendorReservedIdx]));
+        }
+        vendorTagNum++;
+    }
```

Comment:
This makes assumptions about vendor tag usage that are not substantiated by the spec. Would recommend adding a TODO to make this spec compliant. Since this is optional and not going to be emitted by common code in the short term, I don't want to block this PR on that particular TODO as long as an issue is created and its link put in the TODO here.

Reply:
will add TODOs

Reply:
Add issue #9825